x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -384,14 +387,105 @@ public void updateApiKey(
+        for (VersionedApiKeyDoc versionedDoc : targetVersionedDocs) {
+            final String apiKeyId = versionedDoc.id();
+            try {
+                validateForUpdate(apiKeyId, authentication, versionedDoc.doc());
+                final IndexRequest indexRequest = maybeBuildIndexRequest(versionedDoc, authentication, request, userRoleDescriptors);
+                final boolean isNoop = indexRequest == null;
+                if (isNoop) {
+                    logger.debug("Detected noop update request for API key [{}]. Skipping index request", apiKeyId);
+                    responseBuilder.noop(apiKeyId);
+                } else {
+                    bulkRequestBuilder.add(indexRequest);
+                }
+            } catch (Exception ex) {
+                responseBuilder.error(apiKeyId, traceLog("prepare index request for update", ex));
+            }
+        }
+        addErrorsForNotFoundApiKeys(responseBuilder, targetVersionedDocs, request.getIds());
+        if (bulkRequestBuilder.numberOfActions() == 0) {
+            logger.trace("No bulk request execution necessary for API key update");
+            listener.onResponse(responseBuilder.build());
+            return;
+        }
+
+        logger.trace("Executing bulk request to update [{}] API keys", bulkRequestBuilder.numberOfActions());
+        securityIndex.prepareIndexIfNeededThenExecute(
+            ex -> listener.onFailure(traceLog("prepare security index before update", ex)),
+            () -> executeAsyncWithOrigin(
+                client.threadPool().getThreadContext(),
+                SECURITY_ORIGIN,
+                bulkRequestBuilder.setRefreshPolicy(RefreshPolicy.WAIT_UNTIL).request(),
+                ActionListener.<BulkResponse>wrap(
+                    bulkResponse -> buildResponseAndClearCache(bulkResponse, responseBuilder, listener),
+                    ex -> listener.onFailure(traceLog("execute bulk request for update", ex))
+                ),
+                client::bulk
+            )
+        );
+    }
+
+    private void addErrorsForNotFoundApiKeys(
+        final BulkUpdateApiKeyResponse.Builder responseBuilder,
+        final Collection<VersionedApiKeyDoc> foundDocs,
+        final List<String> requestedIds
+    ) {
+        final Set<String> foundIds = foundDocs.stream().map(VersionedApiKeyDoc::id).collect(Collectors.toUnmodifiableSet());
+        for (String id : requestedIds) {
+            if (foundIds.contains(id) == false) {
+                responseBuilder.error(id, new ResourceNotFoundException("no API key owned by requesting user found for ID [" + id + "]"));
+            }
+        }
+    }
+
     // package-private for testing
-    void validateCurrentApiKeyDocForUpdate(String apiKeyId, Authentication authentication, ApiKeyDoc apiKeyDoc) {
+    void validateForUpdate(final String apiKeyId, final Authentication authentication, final ApiKeyDoc apiKeyDoc) {
```

Comment:
Nit rename: the `currentApiKeyDoc` bit didn't seem to add info so removing it for a shorter method name